File: backend/todue/src/main/java/com/ethan/todue/model/Todo.java
package com.ethan.todue.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.Instant;
import java.time.LocalDate;

@Entity
@Table(name = "todos",
    indexes = {
        @Index(name = "idx_user_assigned_date", columnList = "user_id, assigned_date"),
        @Index(name = "idx_user_assigned_completed", columnList = "user_id, assigned_date, is_completed"),
        @Index(name = "idx_user_completed_assigned", columnList = "user_id, is_completed, assigned_date")
    },
    uniqueConstraints = {
        @UniqueConstraint(name = "uk_recurring_instance", columnNames = {"recurring_todo_id", "instance_date"})
    }
)
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Todo {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false, foreignKey = @ForeignKey(name = "fk_todo_user",
        foreignKeyDefinition = "FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE"))
    private User user;

    @NotNull
    @Column(nullable = false, length = 500)
    private String text;

    @NotNull
    @Column(nullable = false)
    private LocalDate assignedDate;

    @NotNull
    @Column(nullable = false)
    private LocalDate instanceDate;

    @Column(nullable = false)
    private Integer position = 0;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "recurring_todo_id", foreignKey = @ForeignKey(name = "fk_todo_recurring",
        foreignKeyDefinition = "FOREIGN KEY (recurring_todo_id) REFERENCES recurring_todos(id) ON DELETE SET NULL"))
    private RecurringTodo recurringTodo;

    @Column(nullable = false, columnDefinition = "BIT(1) DEFAULT 0")
    private Boolean isCompleted = false;

    @Column
    private Instant completedAt;

    @Column(nullable = false, columnDefinition = "BIT(1) DEFAULT 0")
    private Boolean isRolledOver = false;

    @CreationTimestamp
    @Column(nullable = false, updatable = false, columnDefinition = "DATETIME(6) DEFAULT CURRENT_TIMESTAMP(6)")
    private Instant createdAt;

    @UpdateTimestamp
    @Column(nullable = false, columnDefinition = "DATETIME(6) DEFAULT CURRENT_TIMESTAMP(6) ON UPDATE CURRENT_TIMESTAMP(6)")
    private Instant updatedAt;
}

File: mobile/lib/models/todo.dart
class Todo {
  final int? id;
  final String text;
  final String assignedDate; // YYYY-MM-DD format
  final String instanceDate; // YYYY-MM-DD format
  final int position;
  final int? recurringTodoId;
  final bool isCompleted;
  final DateTime? completedAt;
  final bool isRolledOver;
  final bool isVirtual;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Todo({
    this.id,
    required this.text,
    required this.assignedDate,
    required this.instanceDate,
    required this.position,
    this.recurringTodoId,
    required this.isCompleted,
    this.completedAt,
    required this.isRolledOver,
    required this.isVirtual,
    this.createdAt,
    this.updatedAt,
  });

  factory Todo.fromJson(Map<String, dynamic> json) {
    return Todo(
      id: json['id'] as int?,
      text: json['text'] as String,
      assignedDate: json['assignedDate'] as String,
      instanceDate: json['instanceDate'] as String,
      position: json['position'] as int,
      recurringTodoId: json['recurringTodoId'] as int?,
      isCompleted: json['isCompleted'] as bool,
      completedAt: json['completedAt'] != null
          ? DateTime.parse(json['completedAt'] as String)
          : null,
      isRolledOver: json['isRolledOver'] as bool,
      isVirtual: json['isVirtual'] as bool,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'text': text,
      'assignedDate': assignedDate,
      'instanceDate': instanceDate,
      'position': position,
      'recurringTodoId': recurringTodoId,
      'isCompleted': isCompleted,
      'completedAt': completedAt?.toIso8601String(),
      'isRolledOver': isRolledOver,
      'isVirtual': isVirtual,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  Todo copyWith({
    int? id,
    String? text,
    String? assignedDate,
    String? instanceDate,
    int? position,
    int? recurringTodoId,
    bool? isCompleted,
    DateTime? completedAt,
    bool? isRolledOver,
    bool? isVirtual,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Todo(
      id: id ?? this.id,
      text: text ?? this.text,
      assignedDate: assignedDate ?? this.assignedDate,
      instanceDate: instanceDate ?? this.instanceDate,
      position: position ?? this.position,
      recurringTodoId: recurringTodoId ?? this.recurringTodoId,
      isCompleted: isCompleted ?? this.isCompleted,
      completedAt: completedAt ?? this.completedAt,
      isRolledOver: isRolledOver ?? this.isRolledOver,
      isVirtual: isVirtual ?? this.isVirtual,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
